### PR TITLE
Add SCL usage extended tests for S2I images

### PR DESCRIPTION
**Pull request by mfojtik**
Obsoletes: https://github.com/openshift/origin/pull/4029

This might serve as an example for adding more 'test groups'. 

NOTE: The first three commits will be removed as long as https://github.com/openshift/origin/pull/4285 gets merged.

### CODE REVIEW

**Comment by mfojtik** [comment]
@bparees @jhadvig @mnagy PTAL

(only the last commit is interesting).

**Comment by mfojtik** [comment]
@gabemontero this might help you adding new test group.

**Comment by mfojtik** [comment]
[test][extended:s2i]

**Comment by mfojtik** [comment]
@bparees added all S2I images (rhel+centos7)

File: hack/test-extended/s2i/run.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+echo "[INFO] Starting extended tests for s2i ..."
+
+# Run the tests
+pushd ${OS_ROOT}/test/extended >/dev/null
+export KUBECONFIG="${ADMIN_KUBECONFIG}"
+export EXTENDED_TEST_PATH="${OS_ROOT}/test/extended"
+# we don't run in parallel with this suite - do not want different tests tagging the same image in different was at the same time
+ginkgo -progress -stream -v -focus="s2i:" ${OS_OUTPUT_BINPATH}/extended.test
+popd >/dev/null
```
Comment (soltysh) [inline hack/test-extended/s2i/run.sh]: Maybe that will be a stupid question, but I can't resist it. We already have 2 almost identical copies of this script in `hack/test-extended/default/run.sh` and `hack/test-extended/forcepull/run.sh` why? Can't we just create one script to "rule them all" ?

File: test/extended/images/s2i_images.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
+package images
+
+import "fmt"
+
+type tc struct {
+	// The image version string (eg. '27' or '34')
+	Version string
+	// The base OS ('rhel7' or 'centos7')
+	Base string
```
Comment (soltysh) [inline test/extended/images/s2i_images.go]: If that's BaseOS, name the variable accordingly. Base can suggest tons of other things.

File: hack/test-extended/s2i/run.sh (hunk 1 above)
Comment (mfojtik) [inline hack/test-extended/s2i/run.sh]: there is already PR fixing this: https://github.com/openshift/origin/pull/4289

Comment (soltysh) [inline hack/test-extended/s2i/run.sh]: :+1:

File: test/extended/images/scl.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+var _ = g.Describe("s2i: Usage and SCL enablement of the S2I images", func() {
+	defer g.GinkgoRecover()
+	var oc = exutil.NewCLI("s2i-usage", exutil.KubeConfigPath())
+	resolveDockerImageReferences()
+
+	for imageName, tcs := range S2IAllImages() {
+		for _, t := range tcs {
+			g.Describe(fmt.Sprintf("s2i usage of %q", getPodNameForTest(imageName, t)), func() {
+				g.It("should provide usage", func() {
+					g.By("creating a sample pod for that executes the binary directly")
```
Comment (soltysh) [inline test/extended/images/scl.go]: creating a sample pod ~~for~~ that executes the binary directly

Comment (soltysh) [inline test/extended/images/scl.go]: unless you're missing the name after for

File: test/extended/images/scl.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+						Name:  "test",
+						Image: t.DockerImageReference,
+					})
+					oc.KubeFramework().TestContainerOutput(getPodNameForTest(imageName, t), pod, 0, []string{"Sample invocation"})
+				})
+			})
+
+			g.Describe(fmt.Sprintf("rhscl enablement of %q", getPodNameForTest(imageName, t)), func() {
+				g.It("should allow various invocations of binary", func() {
+					g.By(fmt.Sprintf("creating a sample pod for for %q that executes the binary directly", t.DockerImageReference))
```
Comment (soltysh) [inline test/extended/images/scl.go]: creating a sample pod ~~for~~ for %q that executes the binary directly

**Comment by soltysh** [comment]
I like it :+1: 
LGTM, other than some minor spelling errors.

File: test/extended/images/s2i_images.go (hunk 2 above)
Comment (jhadvig) [inline test/extended/images/s2i_images.go]: +1

File: test/extended/builds/sti_env.go [hunk 5]
```diff
@@ -12,6 +12,7 @@ import (
 
 	buildapi "github.com/openshift/origin/pkg/build/api"
 	exutil "github.com/openshift/origin/test/extended/util"
+	kapi "k8s.io/kubernetes/pkg/api"
 )
 
 var _ = g.Describe("default: STI build with .sti/environment file", func() {
```
Comment (jhadvig) [inline test/extended/builds/sti_env.go]: I its not related to this PR, but `s/STI/S2I/` maybe ? :-P

**Comment by jhadvig** [comment]
Besides those tiny nits, LGTM :+1:

**Comment by mfojtik** [comment]
@soltysh @jhadvig thanks! all fixed... now I need to make them pass in Jenkins.

**Comment by jhadvig** [comment]
@mfojtik the force will be with you... always

**Comment by jhadvig** [comment]
[test][extended:all]

**Comment by mfojtik** [comment]
@bparees I think we will have to pre-pull all images in the s2i/run.sh: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/4548/console

this seems jenkins is not able to pull the docker images for RHEL

**Comment by mfojtik** [comment]
[test][extended:s2i]

**Comment by mfojtik** [comment]
@jhadvig you tagged this PR wrong, this is only adding s2i tests, we don't want to execute "all" tests

**Comment by bparees** [comment]
I deleted @jhadvig 's comment, i believe the way the bot works is it aggregates all the requested tests and then runs everything requested anywhere in the PR, it doesn't just look at the last test tag.  at least that's how it worked for v2 extended tests... not sure how it's setup here.

**Comment by jhadvig** [comment]
@mfojtik @bparees sorry, my forgot about that

**Comment by mfojtik** [comment]
[test][extended:s2i]

**Comment by danmcp** [comment]
@mfojtik Needs rebase.

**Comment by mfojtik** [comment]
[test][extended:images]

**Comment by mfojtik** [comment]
@danmcp rebased.

@bparees | | @jhadvig can you review this?

**Comment by bparees** [comment]
lgtm.  [test][extended:core]

**Comment by mfojtik** [comment]
[test][extended:core]

**Comment by mfojtik** [comment]
[test][extended:core]

**Comment by mfojtik** [comment]
[test][extended:core]

**Comment by mfojtik** [comment]
[test][extended:core] on more time

**Comment by mfojtik** [comment]
[test][extended:core(images)]

**Comment by mfojtik** [comment]
[merge] as all scl tests passed

**Comment by mfojtik** [comment]
[merge]

**Pull request merged**